In-Place Pod Vertical Scaling feature

vinaykul opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:
This PR implements API changes defined in the [In-Place Pod Vertical Scaling KEP](https://github.com/kubernetes/enhancements/blob/8bb838caac1251be03785c8974ad72a3e73af629/keps/sig-node/20181106-in-place-update-of-pod-resources.md) , the CRI-API changes defined in [Kubelet CRI KEP](https://github.com/kubernetes/enhancements/blob/8bb838caac1251be03785c8974ad72a3e73af629/keps/sig-node/20191025-kubelet-container-resources-cri-api-changes.md) , and the core implementation of In-Place Pod Vertical Scaling feature that allows users or entities such as VPA to request pod resize without restarting containers.

To summarize changes from the two KEPs referenced above, this PR brings the following:

- Adds new fields named ResourcesAllocated and ResizePolicy to Pod's Container type, and Resources field to ContainerStatus type. It also adds a new admission controller that limits the ability to change ResourcesAllocated field to system-node account. This API change enables users to scale a Pod's CPU and memory resources up or down.

- The CRI change enables runtimes to report currently applied resource limits for a container, and adds support for Windows containers.

The API change code had an initial review by @thockin , and the core implementation has been reviewed by @dashpole and revised. See: https://github.com/vinaykul/kubernetes/pull/1

We are now at a point where we can use further review from the community.

**Alpha Graduation Criteria Status**
- API change and core functionality implemented.
- Resource quota, Limit ranger, scheduler changes, support for kubectl describe is in.
- Resize policy has been implemented.
- Unit tests for the core feature changes are in place.
- E2E tests: Just started working on it.
- Manually tested the at-least following E2E scenarios using kubectl patch:
  - Single container (guaranteed, burstable) Pod, increase resources: CPU + Memory, CPU only, Memory only.
  - Single container (guaranteed, burstable) Pod, decrease resources: CPU + Memory, CPU only, Memory only.
  - Three containers (guaranteed, burstable) Pod, increase CPU of c1, decrease c2, no change for c3, decrease memory of c1, no change for c2 mem, increase memory of c3.
  - Restart kubelet while pod resize is in progress.
  - Inject random failure in dockershim response to UpdateContainerResources CRI API.
  - Dockershim does not return Resources information in ContainerStatus API.
  - Three containers (guaranteed) with RestartContainer resize policy for c2 memory - ensure that c2 is restarted only when c2 memory is resized, regardless of c2 cpu resize.

**Which issue(s) this PR fixes**: 
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #9043

**Special notes for your reviewer**: 

1. The commit named 'InPlace Pod Vertical Scaling feature - API change' implements the API change for In-Place Pod Vertical Scaling.
1. The commit named 'InPlace Pod Vertical Scaling feature - Kubelet CRI change' implements the CRI changes to allow runtimes to report resource limits.
1. The commits named 'InPlace Pod Vertical Scaling feature - core implementation' contains the core kubelet changes that implements the feature, and changes to scheduler to use ResourcesAllocated field, and apiserver to support resource-quota, limit-ranger, and kubectl descibe outputs. Please look at the change to switch-over to ResourcesAllocated, and the resource-quota, limit-ranger changes closely - I have not had a lot of play-time with those.


**Does this PR introduce a user-facing change?**: Yes
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
1. A new field named ResourcesAllocated (type v1.ResourceList) is added to Container struct.
2. A new field named ResizePolicy is added to Container struct.
3. A new field named Resources (type v1.ResourceRequirements) is added to ContainerStatus struct.
4. A new admission controller named 'PodResourceAllocation' is introduced to limit access to modification of ResourcesAllocated field.

e.g: 
root@skibum:~/vInPlacePodVerticalScaling# cat ~/YML/1pod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: 1pod
spec:
  containers:
  - name: stress
    image: skiibum/ubuntu-stress:18.10
    resources:
      limits:
        cpu: "1"
        memory: "1Gi"
      requests:
        cpu: "1"
        memory: "1Gi"
    resizePolicy:
    - resourceName: cpu
      policy: NoRestart
    - resourceName: memory
      policy: RestartContainer
```
The resources can be patched via API or via kubectl as illustrated below:
```
kubectl patch pod 1pod --patch '{"spec":{"containers":[{"name":"stress", "resources":{"requests":{"cpu":"1500m"}, "limits":{"cpu":"1500m"}}}]}}'
```
kubectl describe pod now displays resources allocated to the pod's containers as shown below:
```
root@skibum:~/vInPlacePodVerticalScaling# ./cluster/kubectl.sh describe pod 1pod
Name:         1pod
Namespace:    default
Priority:     0
Node:         127.0.0.1/127.0.0.1
Start Time:   Tue, 30 Jun 2020 14:58:02 -0700
Labels:       <none>
Annotations:  <none>
Status:       Running
IP:           172.18.0.3
IPs:
  IP:  172.18.0.3
Containers:
  stress:
    Container ID:   docker://481892996beb94ee0e0f90de3911f683419ea9d832f765ae762ce893af16feec
    Image:          skiibum/ubuntu-stress:18.10
    Image ID:       docker-pullable://skiibum/ubuntu-stress@sha256:161c50ff4ccc79744c29ea0f337652ce2854ce827914160c7043646b1f8988ba
    Port:           <none>
    Host Port:      <none>
    State:          Running
      Started:      Tue, 30 Jun 2020 14:58:03 -0700
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1500m
      memory:  1Gi
    Requests:
      cpu:     1500m
      memory:  1Gi
    Allocations:
      cpu:        1500m
      memory:     1Gi
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-lvrwt (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  default-token-lvrwt:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-lvrwt
    Optional:    false
QoS Class:       Guaranteed
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type    Reason     Age   From                Message
  ----    ------     ----  ----                -------
  Normal  Scheduled  73s   default-scheduler   Successfully assigned default/1pod to 127.0.0.1
  Normal  Pulled     73s   kubelet, 127.0.0.1  Container image "skiibum/ubuntu-stress:18.10" already present on machine
  Normal  Created    73s   kubelet, 127.0.0.1  Created container stress
  Normal  Started    73s   kubelet, 127.0.0.1  Started container stress
root@skibum:~/vInPlacePodVerticalScaling#
```
Additionally, kubectl describe node displays total node resource allocations and CPU and memory resource allocations for all pods on the node as below:
```
root@skibum:~/vInPlacePodVerticalScaling# ./cluster/kubectl.sh describe node 127.0.0.1
Name:               127.0.0.1
Roles:              <none>
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/os=linux
                    kubernetes.io/arch=amd64
                    kubernetes.io/hostname=127.0.0.1
                    kubernetes.io/os=linux
Annotations:        node.alpha.kubernetes.io/ttl: 0
                    volumes.kubernetes.io/controller-managed-attach-detach: true
CreationTimestamp:  Tue, 30 Jun 2020 14:52:23 -0700
Taints:             <none>
Unschedulable:      false
Lease:
  HolderIdentity:  127.0.0.1
  AcquireTime:     <unset>
  RenewTime:       Tue, 30 Jun 2020 15:04:13 -0700
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  MemoryPressure   False   Tue, 30 Jun 2020 15:02:35 -0700   Tue, 30 Jun 2020 14:52:23 -0700   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Tue, 30 Jun 2020 15:02:35 -0700   Tue, 30 Jun 2020 14:52:23 -0700   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Tue, 30 Jun 2020 15:02:35 -0700   Tue, 30 Jun 2020 14:52:23 -0700   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Tue, 30 Jun 2020 15:02:35 -0700   Tue, 30 Jun 2020 14:52:33 -0700   KubeletReady                 kubelet is posting ready status. AppArmor enabled
Addresses:
  InternalIP:  127.0.0.1
  Hostname:    127.0.0.1
Capacity:
  cpu:                16
  ephemeral-storage:  3807464952Ki
  hugepages-1Gi:      0
  hugepages-2Mi:      0
  memory:             32935964Ki
  pods:               110
Allocatable:
  cpu:                16
  ephemeral-storage:  3508959693954
  hugepages-1Gi:      0
  hugepages-2Mi:      0
  memory:             32833564Ki
  pods:               110
System Info:
  Machine ID:                 91a8ee87a44815729b015c5a5af38563
  System UUID:                49434D53-0200-2500-9072-002590726C8A
  Boot ID:                    5f7a62a7-8d6d-49a5-950b-b89a5a7561c0
  Kernel Version:             4.4.0-184-generic
  OS Image:                   Ubuntu 16.04.6 LTS
  Operating System:           linux
  Architecture:               amd64
  Container Runtime Version:  docker://19.3.8
  Kubelet Version:            v1.19.0-beta.2.529+426d0bbcb13489-dirty
  Kube-Proxy Version:         v1.19.0-beta.2.529+426d0bbcb13489-dirty
Non-terminated Pods:          (2 in total)
  Namespace                   Name                         CPU Requests  CPU Limits  CPU Allocations  Memory Requests  Memory Limits  Memory Allocations  AGE
  ---------                   ----                         ------------  ----------  ---------------  ---------------  -------------  ------------------  ---
  default                     1pod                         1500m (9%)    1500m (9%)  1500m (9%)       1Gi (3%)         1Gi (3%)       1Gi (3%)            6m13s
  kube-system                 kube-dns-5949897dd5-cjqdl    260m (1%)     0 (0%)      260m (1%)        110Mi (0%)       170Mi (0%)     110Mi (0%)          11m
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests     Limits       Allocations
  --------           --------     ------       -----------
  cpu                1760m (11%)  1500m (9%)   1760m (11%)
  memory             1134Mi (3%)  1194Mi (3%)  1134Mi (3%)
  ephemeral-storage  0 (0%)       0 (0%)       0 (0%)
  hugepages-1Gi      0 (0%)       0 (0%)
  hugepages-2Mi      0 (0%)       0 (0%)
Events:
  Type    Reason                   Age                From                   Message
  ----    ------                   ----               ----                   -------
  Normal  Starting                 11m                kubelet, 127.0.0.1     Starting kubelet.
  Normal  NodeHasSufficientMemory  11m (x2 over 11m)  kubelet, 127.0.0.1     Node 127.0.0.1 status is now: NodeHasSufficientMemory
  Normal  NodeHasNoDiskPressure    11m (x2 over 11m)  kubelet, 127.0.0.1     Node 127.0.0.1 status is now: NodeHasNoDiskPressure
  Normal  NodeHasSufficientPID     11m (x2 over 11m)  kubelet, 127.0.0.1     Node 127.0.0.1 status is now: NodeHasSufficientPID
  Normal  NodeAllocatableEnforced  11m                kubelet, 127.0.0.1     Updated Node Allocatable limit across pods
  Normal  Starting                 11m                kube-proxy, 127.0.0.1  Starting kube-proxy.
  Normal  NodeReady                11m                kubelet, 127.0.0.1     Node 127.0.0.1 status is now: NodeReady
root@skibum:~/vInPlacePodVerticalScaling#
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]:
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
In-Place Update of Pod Resources KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Container Resources CRI API Changes for Pod Vertical Scaling KEP:
https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20191025-kubelet-container-resources-cri-api-changes.md
```

Commit:
InPlace Pod Vertical Scaling feature - API change
1. Add ResourcesAllocated and ResizePolicy fields to Container struct.
2. Add Resources field to ContainerStatus struct.
3. Add a new admission controller named PodResourceAllocation for setting defaults and validation.
4. Address code-review items:
   a. Clarify meaning of container resize policy that it is a guidance from user, not a guarantee by k8s.
   b. Handle updates from clients unaware of the new fields in pod spec.
   c. Use DropDisabledPodFields to avoid feature-gate checks in validation where possible.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Commit:
InPlace Pod Vertical Scaling feature - generated code from API change

Commit:
InPlace Pod Vertical Scaling feature - Kubelet CRI change
1. Add ContainerResources message that supports both Linux and Windows resources.
2. Add ContainerResources to ContainerStatus CRI API.
3. Modify UpdateContainerResources CRI API to use ContainerResources.
4. Implement handling ContainerStatus response for runtimes that don't support resources query.
KEP: /enhancements/keps/sig-node/20191025-kubelet-container-resources-cri-api-changes.md

Commit:
InPlace Pod Vertical Scaling feature - core implementation
1. Handle patching of ResourcesAllocated fields of a resized pod in Kubelet's syncPod routine.
2. Handle container limits updation in runtime manager's SyncPod routine.
3. Add new pods after testing for admissibility using ResourcesAllocated field values.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Commit:
InPlace Pod Vertical Scaling feature - review fixes to core implementation
1. Generate CPU Request API status information from runtime's cpu.shares report.
2. Rename pod cgrpoup config get/set functions for functional clarity, and simplify update container resources function.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Commit:
InPlace Pod Vertical Scaling feature - core implementation
1. Switch to using container.ResourcesAllocated instead of container.Resources.Requests for resource allocation/usage calculations.
2. Add support for ResourcesAllocated to kubectl describe with pod and node resources.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Commit:
InPlace Pod Vertical Scaling feature - core implementation
1. Enable resource quota and limit ranger for pod resource changes.
2. In kubelet, do apiserver ResourcesAllocated patch under mutex.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

vinaykul commented:
/assign @dashpole @thockin @liggitt @derekwaynecarr @dchen1107 @ahg-g @PatrickLang @yujuhong

neolit123 commented on cmd/kubeadm/app/features/features.go:
```diff
@@ -32,12 +32,15 @@ const (
-	PublicKeysECDSA: {FeatureSpec: featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha}},
+	IPv6DualStack:             {FeatureSpec: featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha}},
+	PublicKeysECDSA:           {FeatureSpec: featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha}},
+	InPlacePodVerticalScaling: {FeatureSpec: featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha}},
```
you can omit the cmd/kubeadm changes from this PR.
InitFeatureGates contains mostly kubeadm specific features (and not k8s features). it is still possible for users to configure `--feature-gates=InPlacePodVerticalScaling`, for e.g. the kubelet.

vinaykul replied:
Looking back into history, gates such as Auditing, DynamicKubelet, CoreDNS have lived here. Weren't those k8s features being introduced?

neolit123 replied:
yes, but those features also required changes in the kubeadm logic, similarly to the existing `IPv6DualStack`.

`InPlacePodVerticalScaling` is not required to be present to this list as it is a scalability feature that users can pipe trough kubeadm configuration to the components that need it, such as the kubelet.

vinaykul replied:
thanks for the explanation. i've backed out that change in commit https://github.com/kubernetes/kubernetes/pull/92127/commits/113cf1c2ccfb154fd5a5e49de2a8d5f8bd3c4c6f

neolit123 replied:
thank you.

/remove-area kubeadm
/remove-sig cluster-lifecycle

dchen1107 commented:
/ok-to-test

Commit:
InPlace Pod Vertical Scaling feature - test fixes
1. Fix issues found by verify, typecheck, dependencies tests.
2. Fix error message during validation when InPlacePodVerticalScaling feature-gate is off.
3. Backout feature-gate change in kubeadm InitFeatureGates list.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

vinaykul commented:
Odd. Log says [16/16 passed](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92127/pull-kubernetes-verify/1273368578225082370), and my fixes passed locally.

/retest

vinaykul commented:
After investigating e2e-kind-ipv6 test, failures look unrelated to my change. The AdmissionWebhook test passes on my local cluster.

/test pull-kubernetes-e2e-kind-ipv6

dashpole commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -604,6 +632,64 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
+				if container.Resources.Limits == nil || len(pod.Status.ContainerStatuses) == 0 {
+					continue
+				}
+				apiContainerStatus, exists := podutil.GetContainerStatus(pod.Status.ContainerStatuses, container.Name)
```
ideally, we shouldn't depend on the API pod status here.  We should just reconcile the observed (kubecontainer) container status to the resources allocated.

vinaykul replied:
this is to support the transition period where runtime has not implemented reporting Resources in ContainerStatus API response.

apiContainerStatus is populated based on the Spec resources upon successful container creation (initial pod running), and if we get this from CRI, that supercedes spec info. For resizes, we use spec info only if CRI does not yet support it, and updateContainerResources was successful.

dashpole replied:
So essentially you are saying: `v1 pod spec + kubecontainer status = v1 pod status`?  The main thing I am concerned about is that pod.Status.Container[i]... is usually the one stored in the API Server, which can be stale.  I'd prefer only depending on pod.Spec and kubecontainer status if possible.  Could we instead make a helper function for some of the api status generation logic here?  Or we could explictly regenerate the "new" api status to ensure we have consistent behavior.

vinaykul replied:
I had to do this because kubeContainerStatus may be insufficient, and pod.Spec cannot be used without caveats. There are two cases that we have to handle:
1. Runtime has been updated for CRI API change, and returns latest cgroup config in ContainerStatus.Resources,
1. Runtime has not been yet been updated to populate Resources.

Case 1 is easy, we always use cgroup config returned by runtime when we call GetPodStatus after UpdateContainerResources (for newly started container, PLEG is already calling GetPodStatus to query runtime). In this case, kubelet_pods uses runtime supplied config in kubeContainerStatus.Resources to generate apiContainerStatus.

Case 2 is where things get interesting, because we get nothing back from runtime. So we have to go with what we last know - which is what we put into apiContainerStatus. 
- For new container. the container-state transition to running means new container was started, and we trust that the cgroup config passed to the runtime in startContainer has been applied, and we use Resources from Spec.
- For resize, when updateContainerResources is successful, we trust that runtime applied the cgroup config we passed in update, and set that info in pod-cache, which gets consumed when apiStatus is next generated.

So, if we manage what we put into apiContainerStatuses carefully, then we should be ok and I documented in the comments where apiStatus is updated in [kubelet_pods.go](https://github.com/kubernetes/kubernetes/pull/92127/files#diff-10055ae93a8699af13ceba0482fc43c3R1579) 

Now, if the information in apiContainerStatus does get stale or is somehow lost, what's the worst that can happen? We end up executing the update code, and will result in an unnecessary call to updateContainerResources telling runtime to apply config that has already been applied. At some point, we should get rid of case 2 (not sure about the process - perhaps force runtimes to update by adding a gating test if we have such a thing?).

dashpole replied:
The CRI API is supposed to require a set of behaviors of container runtimes.  I agree that we shouldn't require them to add the new function _immediately_, but we should require it eventually, even if they just return exactly what we created the container with...

I wonder if just using the in-place-update feature-gate is the right mechanism?  I.E. Alpha is the opt-in period for runtimes, Beta is when most should have it implemented, but can opt-out if they haven't, and GA is when it is required?  It feels like we are re-implementing the feature graduation process with fall-backs...

vinaykul replied:
Feature-gate alone may not be sufficient to raise awareness. I have to do some research on the process, but I'm hoping that besides socializing the updated ContainerStatus API and updateContainerResources APIs with runtime maintainers, we have some tests that runtimes have to pass in order to work with changes like this.
I think it is reasonable to give at-least one release-cycle after GA before such test becomes mandatory. That should give runtimes enough runway to plan, implement, test ..

dashpole replied:
Ok.  I'm fine with the current implementation for now.  Maybe as a follow-up, you can try and make the 'code to be removed' easier to rip out at a later date, since it is always hard to tell what is safe to remove vs dangerous in later releases.

vinaykul replied:
I'll work on a patch and test it out, and create an issue to track it and apply the patch post-GA of this feature. Let's iron out the details of the plan tomorrow.

dashpole commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -604,6 +632,64 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
+				apiContainerStatus, exists := podutil.GetContainerStatus(pod.Status.ContainerStatuses, container.Name)
+				if !exists || apiContainerStatus.State.Running == nil ||
+					containerStatus.State != kubecontainer.ContainerStateRunning ||
+					containerStatus.ID.String() != apiContainerStatus.ContainerID ||
```
Why does the ID need to be the same?  Is this protecting against resizes during container restarts or something?

vinaykul replied:
Yes. I ran into this with RestartContainer policy where we would enter the resize operation code while the container is already being restarted, and it would get picked for spurious/duplicate kill+starts.

dashpole replied:
A comment explaining this would be helpful.

vinaykul replied:
Added comments. I also changed the entry condition from Resources.Limits to Resources as we need to execute this when only CPU request changes. Please take a look.

dashpole replied:
looks good

dashpole commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -604,6 +632,64 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
+						container: &pod.Spec.Containers[idx],
+						message:   fmt.Sprintf("Container %s resize requires restart", container.Name),
+					}
+					changes.ContainersToRestart = append(changes.ContainersToRestart, idx)
```
can we just use ContainersToStart instead of creating a new ContainersToRestart?  I would hope ContainersToStart would start the container with its new resource requests.

vinaykul replied:
If the resize results in net-increase of pod resources, we need to update the pod-level cgroup values before starting the container. (Step 8 in SyncPod). We also need to apply the changes to any other containers that are being resized without restart so that we don't end up in a situation where the container cgroup values ever exceed the new pod-level cgroup values. I found adding ContainersToRestart invoked at the very end out resize operation pipeline as the cleanest way to achieve this.

vinaykul replied:
Actually, it is possible to get rid of ContainersToRestart if we switched resize step 8 with step 7. Functionally it should work just the same, I think.

However, if we cannot distinguish start from restart, then we would do some amount of possibly throw-away work (ResourceConfigForPod, get-current-pod-cg limit for cpu, mem) when no pod-level cgroup update is needed. If we do this for every new single container, we may see a performance regression.

Thoughts?

dashpole replied:
I generally prefer starting with the simpler implementation and optimizing only if needed.

vinaykul replied:
ok, i'll make this change. we can optimize if there is a perf-regression.

vinaykul replied:
fixed. added a flag to distinguish restart from new container starts, so we shouldn't see a regression.

dashpole commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -385,6 +391,18 @@ type containerToKillInfo struct {
+	// Status of the runtime container
+	kubeContainerStatus *kubecontainer.ContainerStatus
+	// Resize requires restart
+	restart *bool
```
It looks like restart==true means "skip the update".  Can we just not add the container to the list of containers to update instead of adding this field?

vinaykul replied:
Let me see if I can refactor this -- in effect undo append/prepend into ContainersToUpdate array if memory resize can be done without restart, but CPU comes up as restart.

vinaykul replied:
i was able to fix this quite easily, it will be fixed in the next commit.

dashpole commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -385,6 +391,18 @@ type containerToKillInfo struct {
+// containerToUpdateInfo contains necessary information to update a container's resources.
+type containerToUpdateInfo struct {
+	// Value of the index of the specific container in pod.Spec.Containers.
+	specIndex int
```
prefer storing a *v1.Container instead of the index.

vinaykul replied:
fixed.

dashpole commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -829,9 +920,158 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
+				}
+			}
+			if len(podContainerChanges.ContainersToUpdate[rName]) > 0 {
+				if err = m.updatePodContainerResources(pod, podStatus, rName, podContainerChanges.ContainersToUpdate[rName]); err != nil {
```
Why do we need to separate the updatePodContainerResources calls for each resource?

vinaykul replied:
We may be handling a resize where we are doing a net-increase of say memory, while doing a net-decrease of CPU.

dashpole replied:
Got it, thanks

dashpole commented on pkg/kubelet/cm/cgroup_manager_linux.go:
```diff
@@ -777,3 +778,82 @@ func (m *cgroupManagerImpl) GetResourceStats(name CgroupName) (*ResourceStats, e
+
+// Get the cpu quota, cpu period, and cpu shares applied to the cgroup
+func (m *cgroupManagerImpl) GetCgroupCpuConfig(name CgroupName) (int64, uint64, uint64, error) {
+	cgroupPaths := m.buildCgroupPaths(name)
```
I'd encourage you to take a look at https://github.com/opencontainers/runc/blob/master/libcontainer/cgroups/fscommon/utils.go, which handles some of the additional corner cases that comes with reading cgroup files.

vinaykul replied:
Yes, I had gone through this. But I missed TrimSpace check between GetCgroupParamString +  ParseInt below.

vinaykul replied:
GetCgroupParamString has Trimspace check, and max/ParseUint is not applicable here since we are looking for -1, a valid value in this case. Is there something else I missed?

dashpole replied:
Nothing missed necessarily.  IMO using the helper functions they provide seems easier than reimplementing them ourselves.  But its up to you.

vinaykul replied:
Oh yea.. these are merely wrapper functions providing a uniform looking Getter/Setter interface and they are consuming those helpers from ```import github.com/opencontainers/runc/libcontainer/cgroups/fscommon``` . These helpers came in recently i think .. don't remember them being around when I implemented the first prototype.

vinaykul commented on pkg/kubelet/kubelet.go:
```diff
@@ -1637,6 +1648,59 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
+		var resourcesPatchData string
+		for rName, rQuantity := range container.Resources.Requests {
+			container.ResourcesAllocated[rName] = rQuantity
+			resourcesPatchData += fmt.Sprintf(`"%s":"%s",`, rName, rQuantity.String())
```
yes, strategicpatch idea is better. i'll fix it.

vinaykul replied:
fixed in rebased commit.

vinaykul commented on pkg/kubelet/kubelet.go:
```diff
@@ -1637,6 +1648,59 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
+		resourcesPatchData = strings.TrimRight(resourcesPatchData, ",")
+		containersPatchData += fmt.Sprintf(`{"name":"%s","resourcesAllocated":{%s}},`, container.Name, resourcesPatchData)
+	}
+	kl.podManager.UpdatePod(pod)
```
That's true. As of now, we are dealing with pointer to pod object that is being updated.
To clarify, are you suggesting we deepcopy the pod only for patch generation via strategicpatch, and continue to update 'pod' argument, and remove the UpdatePod? The pod object is consumed further down in SyncPod for updateContainerResources. It believe it will see another update from the apiserver patch which is inconsequential for CRI update, I'll verify it.

dashpole replied:
I'd primarily like to be as explicit as possible in any updates we perform to the pod object.  I've had to debug mysterious "why did my pod object randomly change" kubelet bugs.  Given the size of the code base, and the large numbers of readers, it can be hard to figure out what could have possibly updated the pod object in the pod manager.

TBH, i'd prefer doing the deepcopy in the pod manager itself so that updates by readers are impossible, but that will likely cause a regression.  I don't want to couple that with your feature.  In lieu of that, lets just make the update as explicit as possible here.:

1. read the pod
2. deepcopy pod to newPod
3. modify newPod
4. patch(diff(pod, newPod)
5. podManager.UpdatePod(newPod)
6. pod = newPod
... continue on like before.

vinaykul replied:
yes, this is a reasonable compromise. i'll do it this way.

vinaykul replied:
fixed.

Commit:
InPlace Pod Vertical Scaling feature - review fixes
1. Use strategicpatch to generate patch data for ResourcesAllocated when handling pod resize.
2. Use ContainersToStart instead of ContainersToRestart in SyncPod.
3. Simplify use of ContainersToUpdate in SyncPod.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

vinaykul commented on pkg/kubelet/container/helpers.go:
```diff
@@ -102,6 +104,16 @@ func HashContainer(container *v1.Container) uint64 {
 	// Omit nil or empty field when calculating hash value
 	// Please see https://github.com/kubernetes/kubernetes/issues/53644
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		cResources := container.Resources
```
fixed in rebased commit.

vinaykul commented:
```curl -sSL https://kind.sigs.k8s.io/dl/latest/linux-amd64.tgz``` works from my VM. Probably a transient error.

/test pull-kubernetes-e2e-kind

vinaykul commented:
/retest

Commit:
InPlace Pod Vertical Scaling feature - review fixes, initial e2e tests
1. Add framework code for e2e testing, and example e2e tests.
2. Do container resource update based on difference between requests as well as limits.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

vinaykul commented:
```etcd already running at http://127.0.0.1:2379``` appears to be a test conflict.

/retest

vinaykul commented:
/test pull-kubernetes-integration

vinaykul commented:
/assign timothysc andrewsykim neolit123 oomichi Random-Liu tallclair sig-cli-maintainers

alejandrox1 commented on test/e2e/common/pod_resize.go:
```diff
@@ -0,0 +1,437 @@
+limitations under the License.
+*/
+
+package common
```
nit from the testing side: seems that all of this file would be used for a sig node-spnsored feature.
I think it would be better to have this code in a directory that is specific to sig node.
`/test/e2e/common` should have code that would actually be common to other e2e tests.

Could we move it somewhere else?

vinaykul replied:
Thanks for the pointer. I was unsure of this location as well, but found tests for features like empty_dir, runtimeclass etc here. This is a node sponsored feature, but it has a core API change as it adds new fields to Pod Spec & Status, change to kubectl, and a small change to scheduler - so it cuts across the core k8s components. Do you think test/e2e/node is a better location?

alejandrox1 replied:
My initial inclination would be that test/e2e/node would be a bit better (my inclination could be wrong).
Im thinking about the case in which something needs to be done with the test (enhancement or bug fix), what SIG should we ping?

vinaykul replied:
It would be SIG-Node, one of the people listed in the OWNERS file would likely be reviewing the e2e test according to my discussion with Dawn during the weekly sig-node meeting a couple of weeks ago.

I looked more closely and saw the runtimeclass feature e2e test living there, this feature is very congruent to runtimeclass. It looks like test/e2e/node is indeed the right place. I can reconfirm with them.

vinaykul commented:
/assign vpickard

alejandrox1 commented on test/e2e/framework/util.go:
```diff
@@ -506,6 +506,17 @@ func LookForStringInPodExec(ns, podName string, command []string, expectedString
 
+// LookForStringInPodExecToContainer looks for the given string in the output of a command
+// executed in a specific pod container.
+func LookForStringInPodExecToContainer(ns, podName, containerName string, command []string, expectedString string, timeout time.Duration) (result string, err error) {
```
another nit: this kind of code seems generally useful.
I think this would be better in https://github.com/kubernetes/kubernetes/tree/master/test/e2e/framework/pod.
`/test/e2e/framework/util.go` should be for functions directly related to the e2e test framework.

vinaykul replied:
Thanks for the review! Actually, the LookForStringInPodExec function [above](https://github.com/kubernetes/kubernetes/pull/92127/files/426d0bbcb13489c50a52dbbf3b5fef24bb2720d6#diff-eb7b79470992813ea1905e96c298b47bL500) should have been the function I added. The comments on that function mention look for a given string in specific container, but the container name seems to have been missed. I added this so as to not disturb existing references to the above function as part of this very large PR if I fixed the above function.

alejandrox1 replied:
aaah I see. And I also just noticed the comment on the above function - the ideal place for these functions would be not in `e2e/framework/util.go` (for the sake of code organization) but moving functions that use `lookFrString` is non-trivial because they depend on a lot of other code.

Thank you for your reply!

vinaykul replied:
Actually, there are just 17 references to LookForStringIn[PodExec|Log] functions.

It has been bothering me that I need to add ```LookForStringInPodExecToContainer``` where it should really have been LookForStringInPodExec that takes container name as parameter, and defaults to first container if name is empty. There are just 9 references to LookForStringInPodExec at this time, so it is a very manageable churn.

I'm happy to send a refactoring PR for the above, and get rid of the need to add LookForStringInPodExecToContainer - I just don't want it to be part of this large PR. 

Please let me know if you'd like me to create an issue and refactor this - I can certainly move them to another more suitable location while I'm at it. The refactor has to go in first. We are now targeting this PR for v1.20, so there is not much rush from me to get this in.

neolit123 commented:
/unassign

vinaykul commented:
After discussions with @sig-cli in today's meeting, the kubectl describe changes in this PR are best deferred to when we get this feature into beta - they are a nice to have, not critical to the feature. I'll update the PR to back-out the kubectl changes along with moving pod_resize.go from test/e2e/common to test/e2e/node

thockin commented on pkg/apis/core/types.go:
```diff
@@ -1966,6 +1966,33 @@ const (
 
+// ContainerResizePolicy specifies user guidance on how container resource resize should be handled.
+// Only one of the following container resize policies may be specified.
+// If none of the following policies is specified, it defaults to NoRestart.
```
The remarks about default behavior are out of place here, and should go on the field not the type.

thockin replied:
Also, is that the safe default?

vinaykul replied:
By safe, do you mean 'wont have any undesirable results'?

thockin replied:
Assume 90% of users use the default.  Are we going to behave correctly for most of them most of the time?  It seems like "Restart" is the more conservative default, and "NoRestart" is more idealistic.

thockin commented on pkg/apis/core/types.go:
```diff
@@ -1966,6 +1966,33 @@ const (
+type ContainerResizePolicy string
+
+// These are the valid container resize policies:
+// NoRestart policy tells Kubelet to call UpdateContainerResources CRI API to resize
```
This needs to be clearer and comments should be grouped with the constants.  E.g. here's my interpretation:

```
// the resources without restarting the container, if possible. This is the default behavior.
// RestartContainer policy tells Kubelet to stop and start the container with when
// new resources are applied. This is needed for legacy applications e.g. java apps
// using -xmxN flag which are unable to use the resized memory without restarting.
const (
	// NoRestart indicates that this container does not require a restart
	// to have a specific resource (e.g. memory) resized.
	NoRestart ContainerResizePolicy = "NoRestart"
	// Restart indicates that this container requires a restart when a
	// specific resource (e.g. memory) is resized.  For example, Java apps
	// which use the `-xmxN` flag would require to be restarted when the
	// memory resource is resized.
	Restart ContainerResizePolicy = "Restart"
)
```


Note I renamed "RestartContainer" since "container" is redundant...

thockin replied:
Do we need things like RestartOnGrow and RestartOnShrink ?  Should this be RestartAlways?

vinaykul replied:
These could be useful, especially RestartOnShrink since reclaiming memory may not be easily accomplished. @derekwaynecarr Should we add these now? I hope it is OK to updated a merged KEP after-the-fact.

thockin replied:
We can always add them later, as long as we leave room for them.  I don't like designing to hypotheticals, so maybe best to wait until we know we need them?

thockin commented on pkg/apis/core/types.go:
```diff
@@ -1966,6 +1966,33 @@ const (
+// ContainerResizePolicy specifies user guidance on how container resource resize should be handled.
+// Only one of the following container resize policies may be specified.
+// If none of the following policies is specified, it defaults to NoRestart.
+type ContainerResizePolicy string
```
Is this really "Policy" or "Requirement" ?

thockin replied:
Naming wise, I feel like this should be ResourceResizeRequirement or ResourceResizePolicy, and the struct below should be ContainerResizePolicy

vinaykul replied:
Thanks, I agree this looks cleaner. I'll update this to ResourceResizePolicy and ContainerResizePolicy, with values of Restart and NoRestart. RestartContainer got left that way because we had a third RestartPod option at one point which we dropped because we couldn't find a good documented use-case or reason to keep it in the IBM proposal (from where the resize policies came in the first place).

thockin commented on pkg/apis/core/types.go:
```diff
@@ -1966,6 +1966,33 @@ const (
+)
+
+// ResizePolicy represents the resource resize policy for a single container.
+type ResizePolicy struct {
```
ResourceList is one of the only places we use `map` in the API, but it fits well.  There's a value to consistency - why is this better as a struct than a map?

vinaykul replied:
@thockin I wanted this to be a map ([KEP discussion](https://github.com/kubernetes/enhancements/pull/686/#discussion_r320026990)) for simplicity (both for user and for code) despite the API conventions recommending [lists of named subobjects preferred over maps](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps) .. cpu and memory are the limited key choices here vs container object where user can choose the name.

We discussed this during [API live review in 2019 KubeCon San Diego](https://www.youtube.com/watch?v=faRARV3C7Fk&feature=youtu.be&t=3467), and we leaned towards staying with the API conventions as it makes future extensibility simple. 

@liggitt Do you think we can revisit this? Can we make an exception here and use maps even though at this point we don't have any reason for the values to take parameters? I haven't really spent a lot of time seeing APIs evolve, but if you feel there are strong reasons for value to be objects instead of string (my preference because it is simple), we can do that. Are we looking at something like [ContainerState](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L2179)?

thockin replied:
If we make the "value" a struct, then a list probably makes more sense.

If we make the "value" a single value, I have a slight lean towards the map, for congruence with the `resources` field

vinaykul replied:
Then we should leave it as a list - it is possible future iterations may want to qualify the behavior a bit more, such as number of failed update retry attempts before restarting.

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2053,6 +2080,12 @@ type Container struct {
 	Resources ResourceRequirements
+	// Node compute resources allocated to the container.
+	// +optional
+	ResourcesAllocated ResourceList
```
Why is this `spec` and not `status` ?

vinaykul replied:
This field was added only so that we have a source-of-truth of actual node resource allocations to the containers and the pod, should the status field be erased or lost - API conventions need status to be reconstructible.

The status field now has a new Resources field (type ResourceRequirements) to reflect that actual config reported by the runtime via extended CRI

thockin replied:
Wouldn't it be reconstructible from the Kubelet?  Being in spec requires all the admission control mess which is a really string smell that something is off, IMO

vinaykul replied:
Short answer - no.

For point-in-time, there is no way to discover memory request today - it is not stored anywhere. We can reconstruct current CPU requests, limits & memory limits from the extended ContainerStatus CRI API (not today) once runtime supports the CRI extension implemented in this PR.

We had discussed the alternative of checkpointing this locally on the node via atomic file-writes but Dawn mentioned they had run into issues with that kind of approach in Borg, and felt it was safest to have PodSpec be the single source-of-truth.

We had a long discussion on this in sig-node meetings before finally settling on adding ResourcesAllocated to PodSpec, and bringing it up during the Live API review last year. For gory details and kubelet restart scenarios, please see https://github.com/kubernetes/enhancements/pull/686#discussion_r297384981

thockin reviewed (commented):
This is a giant PR, so I only started with API things, and I apologize if some have been covered - hard to catch the whole historical state.  I started re-reading the KEP and there are things there that are unclear to me - is it reasonable to send comments on that?  

Unfortunately - commenting on a committed KEP is hard.  I'll open a PR with <<[UNRESOLVED]>> blocks for discussion.  Is that fair?

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -181,6 +181,11 @@ func NodeRules() []rbacv1.PolicyRule {
+
+	// InPlacePodVerticalScaling
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		nodePolicyRules = append(nodePolicyRules, rbacv1helpers.NewRule("patch").Groups(legacyGroup).Resources("pods").RuleOrDie())
```
cc @tallclair on node update permission on pod specs

vinaykul replied:
/assign tallclair

liggitt commented on plugin/pkg/admission/resourcequota/controller.go:
```diff
@@ -506,7 +508,14 @@ func CheckRequest(quotas []corev1.ResourceQuota, a admission.Attributes, evaluat
-			deltaUsage = quota.SubtractWithNonNegativeResult(deltaUsage, prevUsage)
+			if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+				// allow negative usage for pods as pod resources can increase or decrease
+				if a.GetResource().GroupResource() == corev1.Resource("pods") {
+					deltaUsage = quota.Subtract(deltaUsage, prevUsage)
+				}
```
this leaves non-pod resources with deltaUsage uncomputed... need a test case that would have caught this

liggitt replied:
remind me what protects against making requests that 1) reduce pod resource usage and 2) are constructed to fail to persist (e.g. a conflicting resourceVersion) and artificially lowering consumed quota?

vinaykul replied:
Subtract - 🤦‍♂️  .. will fix and add test.

Regarding gaming of resource quota, we decided against using max(Resources.Requests, ResourcesAllocated) as we don't realistically expect a downward resize to stay unenforced by the kubelet for an extended period of time. Kubelet will always accept downward resize typically in the watch latency.

Please review the rationale noted on the KEP discussion for requests: 
- [resource quota without max](https://github.com/kubernetes/enhancements/pull/686/#discussion_r318246172)()
- [kubelet handling resize without max](https://github.com/kubernetes/enhancements/pull/686/#discussion_r318246172)()

Similarly for limits, iirc [we expected CRI to converge](https://github.com/kubernetes/enhancements/pull/686/#discussion_r324806067) to the new desired limits shortly after watch latency.

liggitt commented on plugin/pkg/admission/podresourceallocation/admission.go:
```diff
@@ -0,0 +1,199 @@
+		if !ok {
+			return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+		}
+		if len(pod.Spec.Containers) != len(oldPod.Spec.Containers) {
```
slight preference for only doing the resource preservation when the length is the same, and letting a length change continue on and get the normal validation error about not changing pod content

```go
if len(pod.Spec.Containers) == len(oldPod.Spec.Containers) {
  ... preserve/populate values
}
```

liggitt commented on plugin/pkg/admission/podresourceallocation/admission.go:
```diff
@@ -0,0 +1,199 @@
+		if !ok {
+			return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+		}
+		if len(pod.Spec.Containers) != len(oldPod.Spec.Containers) {
```
what about initcontainers and ephemeralcontainers?

vinaykul replied:
This feature is not supported for init-containers, we called it out as a non-goal. At that time, I was unaware of ephemeral containers so it is not covered in the KEP. But documentation on it suggested setting resources aren't allowed for ECs, so I didn't worry about it here. I haven't tested it, but I think normal validation will error out if user tries to set EC resources.

liggitt commented on plugin/pkg/admission/podresourceallocation/admission.go:
```diff
@@ -0,0 +1,199 @@
+
+// Admit may mutate the pod's ResourcesAllocated or ResizePolicy fields for compatibility with older client versions
+func (p *Plugin) Admit(ctx context.Context, attributes admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		return nil
+	}
```
prefer injecting the feature gates over taking a direct dependency on the global feature gate. as an example, see:

https://github.com/kubernetes/kubernetes/blob/c53ce48d48372c30053a9b67c6d1dee237b5cd69/plugin/pkg/admission/runtimeclass/admission.go#L82-L87

https://github.com/kubernetes/kubernetes/blob/c53ce48d48372c30053a9b67c6d1dee237b5cd69/plugin/pkg/admission/runtimeclass/admission.go#L99-L106

https://github.com/kubernetes/kubernetes/blob/c53ce48d48372c30053a9b67c6d1dee237b5cd69/plugin/pkg/admission/runtimeclass/admission.go#L116-L120

liggitt commented on pkg/api/pod/util.go:
```diff
@@ -444,6 +444,14 @@ func dropDisabledFields(
+	if !utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) && !inPlacePodVerticalScalingInUse(oldPodSpec) {
+		// Drop ResourcesAllocated and ResizePolicy fields. Don't drop updates to Resources field because
+		// template spec Resources field is mutable for certain controllers. Let ValidatePodUpdate handle it.
+		for i := range podSpec.Containers {
```
also initContainers/ephemeralContainers? should this use the VisitContainers helper?

vinaykul replied:
Let me try out init and ephemeral containers and be sure about them. I think initContainers would need this.

liggitt commented on pkg/api/pod/util.go:
```diff
@@ -559,6 +567,22 @@ func overheadInUse(podSpec *api.PodSpec) bool {
+	}
+	var inUse bool
+	VisitContainers(podSpec, AllContainers, func(c *api.Container, containerType ContainerType) bool {
+		if (c.ResourcesAllocated != nil && len(c.ResourcesAllocated) > 0) || len(c.ResizePolicy) > 0 {
```
can simplify to
```suggestion
		if len(c.ResourcesAllocated) > 0 || len(c.ResizePolicy) > 0 {
```

liggitt commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -192,6 +247,12 @@ func ExtractContainerResourceValue(fs *v1.ResourceFieldSelector, container *v1.C
 		return convertResourceMemoryToString(container.Resources.Requests.Memory(), divisor)
 	case "requests.ephemeral-storage":
 		return convertResourceEphemeralStorageToString(container.Resources.Requests.StorageEphemeral(), divisor)
+	case "resourcesAllocated.cpu":
+		return convertResourceCPUToString(container.ResourcesAllocated.Cpu(), divisor)
+	case "resourcesAllocated.memory":
+		return convertResourceMemoryToString(container.ResourcesAllocated.Memory(), divisor)
+	case "resourcesAllocated.ephemeral-storage":
+		return convertResourceEphemeralStorageToString(container.ResourcesAllocated.StorageEphemeral(), divisor)
```
this extends allowed values for the downward API in ways old kubelets won't understand, right?

vinaykul replied:
It looks like older kubelet should fail validation if users try to expose these fields in their containers (a check I need to update to allow these fields).

End result is that pod won't create, so we won't get to the point where kubelet gets bad info - atleast for 1.18 this is true, I didn't try 1.17, but same validation code exists there. Is this an issue? This seems like the right behavior.

vinaykul replied:
Are you concerned about the case where apiserver has been upgraded, but kubelet has yet to be? I that event, the old kubelet will not recognize these new fields that apiserver lets through - it results in container start error. So not great UX - sometimes it might work and sometimes it won't, until upgrade process complete. 

Is there a way we can handle this?

thockin commented:
@vinay PTAL at https://github.com/kubernetes/enhancements/pull/1883

vinaykul commented:
/assign tallclair

vinaykul commented:
Notes from today's SIG-scheduling weekly meeting discussion on this feature:
- Consider extpanding the container.ResourcesAllocated change in scheduler to InitContainers as well to keep everything uniform.
  - admission plugin should set ResourcesAllocated for init containers as well in this case.
  - validation should reject attempt to mutate init container.
- Verify if this works for upgrade

/cc @ahg-g

ahg-g commented on pkg/scheduler/framework/v1alpha1/types.go:
```diff
@@ -549,8 +549,14 @@ func (n *NodeInfo) resetSlicesIfEmpty() {
-		non0CPUReq, non0MemReq := schedutil.GetNonzeroRequests(&c.Resources.Requests)
+		var non0CPUReq, non0MemReq int64
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+			resPtr.Add(c.ResourcesAllocated)
```
will the ResourcesAllocate field be properly set when upgrading from a cluster that doesn't have this feature enabled (or doesn't have this field at all) to one that does? I assume when api server reads from etcd, it will convert the pod to v1 and at that the conversion logic should probably handle this.

/cc @liggitt

liggitt replied:
yes, as long as the scheduler is speaking to an API server that is aware of the field and has the conversion logic, anything done in conversion will apply to pods that were created in an earlier version when they are read from etcd

vinaykul commented:
I'm closing this PR and will start a new one.

For posterity, we decided that previously approved API changes are not what we really wanted, and we are revisiting the KEP as per discussion in PR https://github.com/kubernetes/enhancements/pull/1883

I'm in the process of refactoring this PR with API changes as per the above discussion. I'll close this PR and start a new one.

@alejandrox1 I'll create a separate issue and a PR that fixes LookForStringInPodExec instead of adding LookForStringInPodExecToContainer as discussed in https://github.com/kubernetes/kubernetes/pull/92127#discussion_r447853904 . The usage of LookForStringInPodExec is still very low this is a zero functionality change, so it should be a breeze to review.

The e2e tests that consume this will come with the main PR for this feature.
